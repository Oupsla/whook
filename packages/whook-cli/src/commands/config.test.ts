import initConfigCommand from './config';
import YError from 'yerror';

describe('configCommand', () => {
  const CONFIGS = {
    MYSQL: {
      auth: {
        username: 'root',
      },
      version: '2.1.1',
    },
  };
  const promptArgs = jest.fn();
  const log = jest.fn();

  beforeEach(() => {
    promptArgs.mockReset();
    log.mockReset();
  });

  it('should work with no query at all', async () => {
    promptArgs.mockResolvedValueOnce({
      _: ['config'],
      name: 'MYSQL',
    });

    const configCommand = await initConfigCommand({
      log,
      CONFIGS,
      promptArgs,
    });
    const result = await configCommand();

    expect({
      result,
      promptArgsCalls: promptArgs.mock.calls,
      logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
    }).toMatchSnapshot();
  });

  it('should work with one value', async () => {
    promptArgs.mockResolvedValueOnce({
      _: ['config'],
      name: 'MYSQL',
      query: 'auth.username',
    });

    const configCommand = await initConfigCommand({
      log,
      CONFIGS,
      promptArgs,
    });
    const result = await configCommand();

    expect({
      result,
      promptArgsCalls: promptArgs.mock.calls,
      logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
    }).toMatchSnapshot();
  });

  it('should work with several values', async () => {
    promptArgs.mockResolvedValueOnce({
      _: ['config'],
      name: 'MYSQL',
      query: 'auth.*',
    });

    const configCommand = await initConfigCommand({
      log,
      CONFIGS,
      promptArgs,
    });
    const result = await configCommand();

    expect({
      result,
      promptArgsCalls: promptArgs.mock.calls,
      logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
    }).toMatchSnapshot();
  });

  it('should work with an unexisting config but a default value', async () => {
    promptArgs.mockResolvedValueOnce({
      _: ['config'],
      name: 'DOES_NOT_EXIST',
      default: 'v8',
    });

    const configCommand = await initConfigCommand({
      log,
      CONFIGS,
      promptArgs,
    });
    const result = await configCommand();

    expect({
      result,
      promptArgsCalls: promptArgs.mock.calls,
      logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
    }).toMatchSnapshot();
  });

  it('should work with no result but a default value', async () => {
    promptArgs.mockResolvedValueOnce({
      _: ['config'],
      name: 'MYSQL',
      query: 'nothing_here',
      default: 'v8',
    });

    const configCommand = await initConfigCommand({
      log,
      CONFIGS,
      promptArgs,
    });
    const result = await configCommand();

    expect({
      result,
      promptArgsCalls: promptArgs.mock.calls,
      logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
    }).toMatchSnapshot();
  });

  it('should fail with unexisting config name', async () => {
    promptArgs.mockResolvedValueOnce({
      _: ['config'],
      name: 'DOES_NOT_EXIST',
    });

    const configCommand = await initConfigCommand({
      log,
      CONFIGS,
      promptArgs,
    });

    try {
      await configCommand();
      throw new YError('E_UNEXPEXTED_SUCCESS');
    } catch (err) {
      expect({
        errorCode: err.code,
        errorParams: err.params,
        promptArgsCalls: promptArgs.mock.calls,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      }).toMatchSnapshot();
    }
  });

  it('should fail with no result', async () => {
    promptArgs.mockResolvedValueOnce({
      _: ['config'],
      name: 'MYSQL',
      query: 'nothing_here',
    });

    const configCommand = await initConfigCommand({
      log,
      CONFIGS,
      promptArgs,
    });

    try {
      await configCommand();
      throw new YError('E_UNEXPEXTED_SUCCESS');
    } catch (err) {
      expect({
        errorCode: err.code,
        errorParams: err.params,
        promptArgsCalls: promptArgs.mock.calls,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      }).toMatchSnapshot();
    }
  });
});
