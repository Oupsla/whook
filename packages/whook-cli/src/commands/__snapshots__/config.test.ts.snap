// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configCommand should fail with no result 1`] = `
Object {
  "errorCode": "E_NO_RESULT",
  "errorParams": Array [
    "MYSQL",
    "nothing_here",
  ],
  "logCalls": Array [
    Array [
      "error",
      "Could not find any results for nothing_here",
    ],
  ],
  "promptArgsCalls": Array [
    Array [],
  ],
}
`;

exports[`configCommand should fail with unexisting config name 1`] = `
Object {
  "errorCode": "E_NO_CONFIG",
  "errorParams": Array [
    "DOES_NOT_EXIST",
  ],
  "logCalls": Array [
    Array [
      "error",
      "No config found for DOES_NOT_EXIST",
    ],
  ],
  "promptArgsCalls": Array [
    Array [],
  ],
}
`;

exports[`configCommand should work with an unexisting config but a default value 1`] = `
Object {
  "logCalls": Array [
    Array [
      "error",
      "No config found for DOES_NOT_EXIST",
    ],
    Array [
      "info",
      "undefined",
    ],
  ],
  "promptArgsCalls": Array [
    Array [],
  ],
  "result": undefined,
}
`;

exports[`configCommand should work with no query at all 1`] = `
Object {
  "logCalls": Array [
    Array [
      "info",
      "{\\"auth\\":{\\"username\\":\\"root\\"},\\"version\\":\\"2.1.1\\"}",
    ],
  ],
  "promptArgsCalls": Array [
    Array [],
  ],
  "result": undefined,
}
`;

exports[`configCommand should work with no result but a default value 1`] = `
Object {
  "logCalls": Array [
    Array [
      "error",
      "Could not find any results for nothing_here",
    ],
    Array [
      "info",
      "\\"v8\\"",
    ],
  ],
  "promptArgsCalls": Array [
    Array [],
  ],
  "result": undefined,
}
`;

exports[`configCommand should work with one value 1`] = `
Object {
  "logCalls": Array [
    Array [
      "info",
      "\\"root\\"",
    ],
  ],
  "promptArgsCalls": Array [
    Array [],
  ],
  "result": undefined,
}
`;

exports[`configCommand should work with several values 1`] = `
Object {
  "logCalls": Array [
    Array [
      "info",
      "\\"root\\"",
    ],
  ],
  "promptArgsCalls": Array [
    Array [],
  ],
  "result": undefined,
}
`;
